src/CommerceContext.php

```diff
@@ -0,0 +1,131 @@
+<?php
+
+namespace Frontkom\DrupalBehatDefinitions;
+
+use Behat\Gherkin\Node\TableNode;
+use Drupal\DrupalExtension\Context\RawDrupalContext;
+use Drupal\commerce_product\ProductVariationStorageInterface;
+use Drupal\commerce_promotion\Entity\PromotionInterface;
+
+/**
+ * Class CommerceContext.
+ *
+ * Provide Behat step-definitions for common Commerce functionalities.
+ */
+class CommerceContext extends EntityContextBase {
+
+  /**
+   * Generate coupons.
+   *
+   * @Given coupons:
+   */
+  public function createCoupons(TableNode $nodesTable) {
+    $storage = \Drupal::entityTypeManager()->getStorage('commerce_promotion_coupon');
+    foreach ($nodesTable->getHash() as $nodeHash) {
+      $coupon = (object) $nodeHash;
+      $promotion = $this->getPromotionByName($coupon->promotion);
+
+      $coupon_saved = $this->couponCreate($coupon);
+      $coupon_loaded = $storage->load($coupon_saved->id);
+      $promotion->get('coupons')->appendItem($coupon_loaded);
+      $promotion->save();
+    }
+  }
+
+  /**
+   * Load promotion by name.
+   */
+  public function getPromotionByName($name) {
+    $storage = \Drupal::entityTypeManager()->getStorage('commerce_promotion');
+    $promotions = $storage->loadByProperties(['name' => $name]);
+
+    if (count($promotions) !== 1) {
+      throw new \Exception('Expected 1 promotion with title ' . $name . ' but found ' . count($promotions));
+    }
+
+    return reset($promotions);
+  }
+
+  /**
+   * Create coupon.
+   */
+  public function couponCreate($coupon) {
+    $saved = $this->getDriver()->createEntity('commerce_promotion_coupon', $coupon);
```

Comment:
Will this make them automatically deleted after the scenario?

Reply:
I am now 100% sure that is not the case.

The entities are probably randomly deleted only because they are tied to a promotion deleted in the same cleanup. But there is nothing in our test classes or drupal extension to clean it up. We should do that here

Reply:
Should be as easy as append it to $this->entities though